model.py

```diff
@@ -134,51 +147,98 @@ def forward(self, query, values):
 
         # Weight tying.
-        if kwargs['tie_embed']:
+        if self.xent and kwargs["tie_embed"]:
             self.linear1.weight = self.embedding.weight
 
-    def forward(self, trg_batch, h_encoder):
+    def step(
+        self,
+        token_embedding: Tensor,
+        model_output: Tensor,
+        hidden: Tensor,
+        h_encoder: Tensor,
+    ) -> Tuple[Tensor, Tuple[Tensor, Tensor]]:
+        input_combined = torch.cat([token_embedding, model_output], dim=-1)
+        h_decoder, hidden = self.lstm(input_combined, hidden)
+        attended_output = self.global_attn(h_decoder, h_encoder)
+        return self.dropout(attended_output), hidden
+
+    def forward(self, trg_batch: Tensor, h_encoder: Tensor) -> Tensor:
         trg_embeddings = self.embedding(trg_batch)
 
-        output = torch.zeros([1, trg_batch.shape[1], trg_embeddings.shape[2]])\
-            .to(self.embedding.weight.device)
+        output = torch.zeros([1, trg_batch.shape[1], trg_embeddings.shape[2]]).to(
+            self.embedding.weight.device
+        )
         hidden = None
         outputs = []
         for i in range(trg_embeddings.shape[0]):
-            trg_embed = trg_embeddings[i:i+1,:,:] # t=1 x b x d
-            input_combined = torch.cat([trg_embed, output], dim=-1)
-            h_decoder, hidden = self.lstm(input_combined, hidden)
-            attended_output = self.global_attn(h_decoder, h_encoder)
-            output = self.dropout(attended_output)
+            trg_embed = trg_embeddings[i : i + 1, :, :]  # t=1 x b x d
+            output, hidden = self.step(trg_embed, output, hidden, h_encoder)
             if self.xent:
                 output_projected = self.linear1(output)
                 outputs.append(output_projected)
             else:
                 outputs.append(output)
         return torch.cat(outputs, 0)
 
+    def decode(
+        self, h_encoder: Tensor, max_decoding_len: int, bos_idx: int, eos_idx: int,
+    ) -> List[List[int]]:
+        bos_idx_tensor = torch.LongTensor([bos_idx]).to(self.embedding.weight.device)
+        bos_embed = self.embedding(bos_idx_tensor)
+        batch_size = h_encoder.shape[1]
+        decoded_embeds = [bos_embed.reshape(1, -1).repeat(1, batch_size, 1)]
+        model_out = output = torch.zeros(
+            [1, batch_size, self.embedding.weight.shape[1]]
+        ).to(self.embedding.weight.device)
+        hidden = None
+        decoded_idxs = [np.array([[bos_idx]]).repeat(batch_size, 1)]
+        eos_generated = np.zeros((1, batch_size), dtype=np.bool)
+        while len(decoded_idxs) < max_decoding_len and (eos_generated == 0).any():
+            model_out, hidden = self.step(  ###
+                decoded_embeds[-1], model_out, hidden, h_encoder,
+            )
+            if self.xent:
+                model_sm = self.linear1(model_out)
+                decoded_idxs.append(model_sm.argmax(-1).cpu().numpy())
+                decoded_embeds.append(self.embedding(model_sm.argmax(-1)))
```

Comment:
Instead of calling the embedding function here and above for bos token, can you do it at the beginning of the while loop before calling step?